src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DelayUnassignedAllocation.java

```diff
@@ -0,0 +1,307 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.allocator;
+
+import com.carrotsearch.hppc.cursors.ObjectCursor;
+import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.Lists;
+import org.apache.lucene.util.ArrayUtil;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.routing.MutableShardRouting;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The delayed unassigned allocation allows to control delayed allocation of shards due to nodes leaving the
+ * cluster with the assumption that they might come back, and the opertunity to reduce cluster allocations if
+ * they do.
+ */
+public class DelayUnassignedAllocation extends AbstractComponent {
+
+    public static final String DELAY_ALLOCATION_DURATION = "cluster.routing.allocation.delay_unassigned_allocation.duration";
+    public static final String DELAY_ALLOCATION_NODE_KEY = "cluster.routing.allocation.delay_unassigned_allocation.node_key";
+
+    private final NodeKey nodeKey;
+    private volatile TimeValue duration;
+
+    private final Map<Object, DelayedAllocation> delayedAllocations = new HashMap<>();
+    private final Cache<String, DiscoveryNode> seenNodes = CacheBuilder.newBuilder().expireAfterWrite(5, TimeUnit.MINUTES).build();
+
+    public DelayUnassignedAllocation(Settings settings) {
+        super(settings);
+        this.duration = settings.getAsTime(DELAY_ALLOCATION_DURATION, TimeValue.timeValueMinutes(5));
+        this.nodeKey = NodeKey.valueOf(settings.get(DELAY_ALLOCATION_NODE_KEY, NodeKey.TRANSPORT_ADDRESS.toString()).toUpperCase(Locale.ROOT));
```

Comment:
Yeah, I guess I'm thinking of people deploying without static IPs (certainly not recommended). These most likely won't suffer the same problems that this solution helps though, so it may be a moot point altogether.